proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+
+| Opcode | Type constructor |
+|--------|------------------|
+| -0x41  |  `except_ref`    |
+
+#### value_type
+
+A `varint7` indicating a a `value type` is extended to include `except_ref` as
+encoded above.
+
+#### Other Types
+
+##### except_type
+
+An exception is described by its exception type signature, which corresponds to
+the data fields of the exception.
+
+| Field | Type | Description |
+|-------|------|-------------|
+| `count` | `varuint32` | The number of types in the signature |
+| `type` | `value_type*` | The type of each element in the signature |
+
+
+##### external_kind
+
+A single-byte unsigned integer indicating the kind of definition being imported
+or defined:
+
+* `0` indicating a `Function` [import](Modules.md#imports) or [definition](Modules.md#function-and-code-sections)
+* `1` indicating a `Table` [import](Modules.md#imports) or [definition](Modules.md#table-section)
+* `2` indicating a `Memory` [import](Modules.md#imports) or [definition](Modules.md#linear-memory-section)
+* `3` indicating a `Global` [import](Modules.md#imports) or [definition](Modules.md#global-section)
+* `4` indicating an `Exception` [import](#import-section) or [definition](#exception-sectio)
+
+### Module structure
+
+#### High-level structure
+
+A new `exception` section is introduced and is named `exception`. If included,
+it must appear between the `Export` and `Start` sections of the module.
+
+
+##### Exception section
+
+The `exception` section is the named section 'exception'. The exception section
+declares exception types using exception type signatures.
+
+| Field | Type | Description |
+|-------|------|-------------|
+| count | `varuint32` | count of the number of exceptions to follow |
+| sig | `except_type*` | The type signature of the data fields for each exception |
+
+
+##### Import section
+
+The import section is extended to include exception types by extending an
+`import_entry` as follows:
+
+If the `kind` is `Exception`:
+
+| Field | Type | Description |
+|-------|------|-------------|
+| `sig`  | `except_type` | the type signature of the exception |
+
+##### Export section
+
+The export section is extended to include exception types by extending an
+`export_entry` as follows:
+
+If the `kind` is `Exception`, then the `index` is into the corresponding
+exception in the [exception index space](#exception-index-space).
+
+
+##### Name section
+
+The set of known values for `name_type` of a name section is extended as
+follows:
+
+| Name Type | Code | Description |
+| --------- | ---- | ----------- |
+| [Function](#function-names) | `1` | Assigns names to functions |
+| [Local](#local-names) | `2` | Assigns names to locals in functions |
+| [Exception](#exception-names) | `3` | Assigns names to exception types |
+
+###### Exception names
+
+The exception names subsection is a `name_map` which assigns names to a subset
+of the _exception_ indices from the exception section. (Used for both imports
+and module-defined).
```

Comment:
I can be mistaken, but can there be any debug info attached to not an actual exception object but an exception tag?

Reply:
You could attach a name to an exception tag, which could count as very basic debug info, but in general I'd say no. Runtime info like a stack trace can only be attached to an exception object.

Reply:
Then what can the capability of attaching a name to an exception index be used for?

Reply:
It'd be similar to attaching a name to a function. Without names, Wasm call stacks basically look like "wasm_function_0, wasm_function_1, wasm_function_1, ...." In the same way, names on exceptions would let devtools say something like "Unhandled exception: foo" rather than "Unhandled Exception: my_module.wasm:0" or worse, "Unhandled exception: <wasm exception 0x4ea2df>".

Reply:
What you describe is we attach debug info to exception objects. But if we assign names to exception indices, that doesn't mean individual exception objects can have separate names, no?

Reply:
Correct, this would just be a name for the exception type, not the individual object.